syfertext/tokenizer.py

```diff
@@ -175,18 +206,365 @@ def __call__(self, text: Union[String, str]):
                 )
 
-                # Append the token to the document
-                doc.container.append(token_meta)
+                # if is_space is True that means detected token is composed of only whitespaces
+                # so we dont need to check for prefix, infixes etc.
+                if is_space:
+
+                    # Append the token to the document
+                    doc.container.append(token_meta)
+                else:
+
+                    # Process substring for prefix, infix, suffix and exception cases
+                    span = str(text[pos:None])
+                    doc = self._tokenize(span, token_meta, doc)
 
         return doc
 
+    def _tokenize(self, substring: str, token_meta: TokenMeta, doc: Doc) -> Doc:
+        """ Tokenize each substring formed after spliting affixes and processing 
+            exceptions(special cases). Returns Doc object.
+
+        Args:
+            substring (str) : The substring to tokenize.
+            token_meta (TokenMeta) : The TokenMeta object of original substring
+                before spliting affixes and special cases.
+            doc (Doc): Document object. 
+
+        Returns:    
+            Document with all the TokenMeta objects of every token after spliting 
+            affixes and exceptions.
+        """
+
+        pos = token_meta.start_pos
+        space_after = token_meta.space_after
+        (substring, pos, prefixes, suffixes, infixes, special_tokens,) = self._split_affixes(
+            substring, pos
+        )
+        doc = self._attach_tokens(
+            doc, substring, pos, space_after, prefixes, suffixes, infixes, special_tokens,
+        )
+
+        return doc
+
+    def _split_affixes(
+        self, substring: str, start_pos: int
+    ) -> Tuple[str, int, List[TokenMeta], List[TokenMeta], List[TokenMeta], List[TokenMeta]]:
+        """Process substring for tokenizinf prefixes, infixes, suffixes and exceptions.
+
+        Args:
+            substring (str) : The substring to tokenize.
+            start_pos (int) : The pointer to location of start of substring in text.
+
+        Returns:    
+            substring: The substring to tokenize.
+            start_pos: The pointer to location of start of substring in text.
+            prefixes: The list of prefixes TokenMeta objects.
+            suffixes: The list of suffixes TokenMeta objects.
+            infixes: The list of infixes TokenMeta objects.
+            special_tokens: The list of special tokens TokenMeta objects.
+        """
+        suffixes = []
+        prefixes = []
+        infixes = []
+        special_tokens = []
+        pos = start_pos
+        end_pos = pos
+
+        if substring:
+            is_space = False
+            space_after = False  # for the last token meta space after will be updated explicitely according to the original substring.
+            while self.find_prefix(substring) or self.find_suffix(substring):
```

Comment:
next_affix = ['prefix', 'suffix']

        # Dict holding TokenMeta lists of each affix types(prefix, suffix, infix)
        affixes = defaultdict(list)

        # The first element in the `next_affix` list is 'prefix'.
        # since we should start by finding prefixes, we fix the
        # index i = 0.
        # Start by finding prefixes. 
        i = 1

        # Loop until an empty substring is encountered
        while self.find_prefix(substring) or self.find_suffix(substring):

            if substring in self.exceptions:
                break

            # Get affix type (prefix or suffix) 
            affix_type = next_affix[i%2]
                       
            # Get the `TokenMeta` object of the affix along with updated 
            # substring and start pos pointer after removing the affix.
            token_meta, substring, pos = getattr(self, f'_get_{affix_type}_token_meta')(substring, pos)
            affixes[f"{affix_type}"].append(token_meta)

            # Change the affix type.
            i += 1

Reply:
Yup updated it